### DISCUSSION TITLE
Implementing route based code splitting for a Next.js project

### DISCUSSION THREAD

**Discussion by markusylisiurunen** [vercel/next.js#12778] [Help]
Hi!

I'm currently working with a Next.js project that has a few separated (kind of) areas. The three different areas are:

1. Landing page (+ some subpages)
2. Blog
3. Dashboard

The project structure is the following:

```
- /pages
  - index.js
  - /blog
    - index.js
    - [id].js
  - /dashboard
    - <bunch of dashboard routes w/ subfolders...>
```

The dashboard uses bunch of libraries for things like GraphQL (`@apollo/client`), input validation and so on that are not needed on the other two areas. But as there are so many different paths in the dashboard folder, for example, `@apollo/client`  gets imported so many times (probably like 90%+ of all routes) that Webpack decides that its common enough to include in the common chunk which is loaded on every page (that itself is over 30kB of JS gzipped), including the landing page which does not need it.

I’d like to somehow make Webpack separate the dashboard folder from other stuff so that it creates a common chunk for routes under `/dashboard` which does not get included in blog or marketing pages.

I've tried inspecting the Webpack `SplitChunksPlugin` configuration and tried adding the following cache group to that.

```js
config.optimization.splitChunks.cacheGroups.dashboard = {
  name: "dashboard",
  minChunks: 2,
  priority: 15,
  reuseExistingChunk: true,
  test: /[\\/]node_modules[\\/]/,
  chunks(chunk) {
    return /pages[\\/]dashboard/.test(chunk.name);
  },
};
```

When inspecting with `@next/bundle-analyzer` it seems to work because, for example, `@apollo/client` is now in the `dashboard` chunk and not in anything else.

However, at least in development mode, I get frequent errors when navigating through the dashboard of some missing modules/invalid exports etc. One example is shown below.

<img width="535" alt="Screenshot 2020-05-12 at 12 38 14" src="https://user-images.githubusercontent.com/8409947/81668543-7dab6a80-944d-11ea-822b-0154bbb59cab.png">

Based on these errors, my guess is that some modules (`moment`) get bundled into multiple different chunks and they don't work together.

My question is, does someone know of a good way to do this kind of code splitting with the Webpack's `SplitChunksPlugin`? Any tips what could be wrong?

**Comment by markusylisiurunen** [comment]
This is the default `SplitChunksPlugin` configuration that Next.js uses.

```json
{
  "chunks": "all",
  "cacheGroups": {
    "default": false,
    "vendors": false,
    "framework": {
      "chunks": "all",
      "name": "framework",
      "test": {},
      "priority": 40,
      "enforce": true
    },
    "lib": {
      "priority": 30,
      "minChunks": 1,
      "reuseExistingChunk": true
    },
    "commons": {
      "name": "commons",
      "minChunks": 33,
      "priority": 20
    },
    "shared": {
      "priority": 10,
      "minChunks": 2,
      "reuseExistingChunk": true
    },
    "styles": {
      "name": "styles",
      "test": {},
      "chunks": "all",
      "enforce": true
    }
  },
  "maxInitialRequests": 25,
  "minSize": 20000
}
```

I don't quite understand how this works. What does `{}` test for? Is it somehow populated at runtime after `next.config.js` is run?

What should be the priority for the new `dashboard` cache group?

**Reply by PlopTheReal** [reply]
Hey, 
I'm facing the problem of figuring out how to split things better, have you found anything?

**Comment by callum-gander** [comment]
This is exactly the same problem I'm facing. There are a bunch of very large node modules that just aren't necessary when the user initially visits the landing page and aren't necessary until the user visits the specific page they're necessary for. How can we essentially not bundle those pages and nodule modules until we go to that page and then fetch them, or fetch them in the background after the initial fetch?
